table/block_based_table_reader.cc

```diff
@@ -1987,10 +1989,24 @@ void BlockBasedTableIterator::InitDataBlock() {
         num_file_reads_++;
         // Do not readahead more than kMaxReadaheadSize.
         readahead_size_ = std::min(kMaxReadaheadSize, readahead_size_);
-        table_->get_rep()->file->Prefetch(data_block_handle.offset(),
-                                          readahead_size_);
-        readahead_limit_ = static_cast<size_t>(data_block_handle.offset()
-          + readahead_size_);
+
+        // Discarding the return status of Prefetch calls intentionally, as we
+        // can fallback to reading from disk if Prefetch fails.
+        if (!rep->file->use_direct_io()) {
+          // buffered i/o
+          rep->file->Prefetch(data_block_handle.offset(), readahead_size_);
+          readahead_limit_ =
+              static_cast<size_t>(data_block_handle.offset() + readahead_size_);
+        } else if (!rep->for_compaction) {
+          // direct i/o -- but not for compaction reads.
```

Comment:
Yeah, definitely, in the future. The current problem is that `compaction_readahead_size` interferes with this. When we extend this to compactions we can deprecate `compaction_readahead_size`. Since my current goal is to improve user range reads, I didn't want to touch compactions.